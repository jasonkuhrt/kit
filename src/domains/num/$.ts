/**
 * Numeric types and utilities with branded types for mathematical constraints.
 *
 * Provides branded number types (Positive, Negative, Even, Odd, etc.) with
 * runtime validation, mathematical operations, range types, and specialized
 * numeric domains like Complex, Ratio, and BigInt. Includes type guards,
 * ordering, and equivalence utilities.
 */
export * as Num from './$$.js'
