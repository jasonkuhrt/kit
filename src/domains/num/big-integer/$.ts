export * as BigInteger from './$$.ts'
