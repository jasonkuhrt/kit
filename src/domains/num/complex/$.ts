export * as Complex from './$$.js'
