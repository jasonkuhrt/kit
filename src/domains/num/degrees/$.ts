// Export the namespace containing all Degrees operations
export * as Degrees from './$$.js'
