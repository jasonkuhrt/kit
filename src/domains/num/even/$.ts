// Export the namespace containing all Even operations
export * as Even from './$$.ts'
