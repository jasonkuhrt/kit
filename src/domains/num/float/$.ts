// Export the namespace containing all Float operations
export * as Float from './$$.js'
