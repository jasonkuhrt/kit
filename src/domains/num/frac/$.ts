// Export the namespace containing all Frac operations
export * as Frac from './$$.js'
