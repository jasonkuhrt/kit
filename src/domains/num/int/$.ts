// Export the namespace containing all Int operations
export * as Int from './$$.ts'
