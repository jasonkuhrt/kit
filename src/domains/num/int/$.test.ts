import { expect, test } from 'vitest'
import { Int } from './$.ts'

test('isInt', () => {
  // Valid cases - integers
  expect(Int.is(0)).toBe(true)
  expect(Int.is(1)).toBe(true)
  expect(Int.is(-1)).toBe(true)
  expect(Int.is(100)).toBe(true)
  expect(Int.is(-100)).toBe(true)
  expect(Int.is(Number.MAX_SAFE_INTEGER)).toBe(true)
  expect(Int.is(Number.MIN_SAFE_INTEGER)).toBe(true)

  // Invalid cases - non-integers
  expect(Int.is(1.5)).toBe(false)
  expect(Int.is(-1.5)).toBe(false)
  expect(Int.is(0.1)).toBe(false)
  expect(Int.is(Infinity)).toBe(false)
  expect(Int.is(-Infinity)).toBe(false)
  expect(Int.is(NaN)).toBe(false)
  expect(Int.is('1')).toBe(false)
  expect(Int.is(null)).toBe(false)
  expect(Int.is(undefined)).toBe(false)
})

test('int', () => {
  // Valid cases
  expect(Int.from(0)).toBe(0)
  expect(Int.from(1)).toBe(1)
  expect(Int.from(-1)).toBe(-1)
  expect(Int.from(100)).toBe(100)
  expect(Int.from(-100)).toBe(-100)
  expect(Int.from(Number.MAX_SAFE_INTEGER)).toBe(Number.MAX_SAFE_INTEGER)
  expect(Int.from(Number.MIN_SAFE_INTEGER)).toBe(Number.MIN_SAFE_INTEGER)

  // Invalid cases - should throw
  expect(() => Int.from(1.5)).toThrow('Value must be an integer')
  expect(() => Int.from(-1.5)).toThrow('Value must be an integer')
  expect(() => Int.from(0.1)).toThrow('Value must be an integer')
  expect(() => Int.from(Infinity)).toThrow('Value must be an integer')
  expect(() => Int.from(-Infinity)).toThrow('Value must be an integer')
  expect(() => Int.from(NaN)).toThrow('Value must be an integer')
})

test('tryInt', () => {
  // Valid cases
  expect(Int.tryFrom(0)).toBe(0)
  expect(Int.tryFrom(1)).toBe(1)
  expect(Int.tryFrom(-1)).toBe(-1)
  expect(Int.tryFrom(100)).toBe(100)
  expect(Int.tryFrom(-100)).toBe(-100)
  expect(Int.tryFrom(Number.MAX_SAFE_INTEGER)).toBe(Number.MAX_SAFE_INTEGER)
  expect(Int.tryFrom(Number.MIN_SAFE_INTEGER)).toBe(Number.MIN_SAFE_INTEGER)

  // Invalid cases - should return null
  expect(Int.tryFrom(1.5)).toBe(null)
  expect(Int.tryFrom(-1.5)).toBe(null)
  expect(Int.tryFrom(0.1)).toBe(null)
  expect(Int.tryFrom(Infinity)).toBe(null)
  expect(Int.tryFrom(-Infinity)).toBe(null)
  expect(Int.tryFrom(NaN)).toBe(null)
})
