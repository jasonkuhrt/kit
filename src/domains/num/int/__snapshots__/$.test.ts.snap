// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`from > from(-1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-1
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
-1
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(-1.5) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-1.5
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Value must be an integer, got: -1.5
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(-100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-100
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
-100
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(-9007199254740991) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-9007199254740991
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
-9007199254740991
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(-Infinity) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-Infinity
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Value must be an integer, got: -Infinity
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(0) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
0
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
0
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(0.1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
0.1
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Value must be an integer, got: 0.1
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
1
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(1.5) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1.5
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Value must be an integer, got: 1.5
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
100
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
100
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(9007199254740991) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
9007199254740991
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
9007199254740991
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(Infinity) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
Infinity
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Value must be an integer, got: Infinity
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(NaN) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
NaN
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Value must be an integer, got: NaN
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is('1') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(-1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-1
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(-1.5) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-1.5
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(-100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-100
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(-9007199254740991) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-9007199254740991
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(-Infinity) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-Infinity
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(0) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
0
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(0.1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
0.1
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(1.5) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1.5
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
100
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(9007199254740991) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
9007199254740991
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(Infinity) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
Infinity
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(NaN) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
NaN
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(null) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
null
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(undefined) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
undefined
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(-1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-1
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
-1
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(-1.5) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-1.5
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(-100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-100
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
-100
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(-9007199254740991) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-9007199254740991
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
-9007199254740991
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(-Infinity) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-Infinity
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(0) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
0
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
0
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(0.1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
0.1
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
1
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(1.5) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1.5
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
100
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
100
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(9007199254740991) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
9007199254740991
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
9007199254740991
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(Infinity) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
Infinity
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(NaN) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
NaN
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;
